pixeltable/utils/hf_datasets.py

```diff
@@ -0,0 +1,111 @@
+import datasets
+from typing import Union, Optional, List, Dict, Any
+import pixeltable.type_system as ts
+from pixeltable import exceptions as excs
+import math
+import logging
+
+_logger = logging.getLogger(__name__)
+
+# https://huggingface.co/docs/datasets/v2.17.0/en/package_reference/main_classes#datasets.Value
+_hf_to_pt : Dict[str, ts.ColumnType] = {
+    'uint32' : ts.IntType(nullable=True),
+    'uint64' : ts.IntType(nullable=True),
+    'int32' : ts.IntType(nullable=True),
+    'int64' : ts.IntType(nullable=True),
+    'bool' : ts.BoolType(nullable=True),
+    'float32' : ts.FloatType(nullable=True),
+    'string' : ts.StringType(nullable=True),
+    'timestamp[s]' : ts.TimestampType(nullable=True),
+    'timestamp[ms]' : ts.TimestampType(nullable=True),
+}
+
+def create_table_from_huggingface_dataset(self : 'pixeltable.Client',
+                                            path_str: str,
+                                            dataset : Union['datasets.Dataset', 'datasets.DatasetDict'],
+                                            split_column_name : Optional[str],
+                                            explain_only : bool,
+                                            primary_key: Union[str, List[str]],
+                                            num_retained_versions: int) -> Optional['catalog.InsertableTable']:
+    """ See pixeltable.Client.create_table_from_huggingface_dataset for documentation
+    """
+
+    dataset_dict : Dict[str, datasets.Dataset] = {}
+    if isinstance(dataset, datasets.Dataset):
+        # get split string while removing any slice notation
+        if dataset.split._name is not None:
+            # when loading an hf dataset partially, dataset.split._name is sometimes the form "train[0:1000]"
+            # in those cases, we only only want to keep the "train" part
+            split_name = dataset.split._name.split('[')[0]
+        else:
+            split_name = None
+
+        if split_name is None and split_column_name is not None:
+            print('no split name found in dataset, ignoring')
+            split_column_name = None
+
+        dataset_dict = {split_name:dataset}
+    elif isinstance(dataset, datasets.DatasetDict):
+        for column_name in dataset:
+            dataset_dict[column_name] = dataset[column_name]
+    else:
+        raise excs.Error(f'type(dataset) must be datasets.Dataset or datasets.DatasetDict. Got {type(dataset)}')
```

Comment:
This should maybe be an `assert false`? Since it's implied by the parameter type.

Reply:
This code is akin to user input validation, checking the user met the type contract, as the method is meant to be exposed and used externally.
Assertions tend to look like internal errors.

Reply:
Right, I understand the difference between assertions and exceptions. I was suggesting that this should be an assertion because the type is already enforced by the method signature, so in some sense it is an internal error.

But I guess in python the method signature typecheck isn't actually enforced by the runtime so it's kind of a grey area. Someone *could* pass something in that violates the method contract and that would be a user error, not an internal error.

A matcher would be a far better pattern here than if/elif/else but we can't use python 3.10 features sadly :(